Port 2 more OPAL config files to OAC

jsquyres opened the pull request:
- oac_summary.m4
- oac_setup_sphinx.m4

See individual commit messages for details.

jsquyres commented:
@bwbarrett What do you think of this?  It is motivated by me (finally) working on Sphinx support for PRTE and PMIX: it was easy enough to convert most of `OPAL_SETUP_SPHINX` to `OAC_SETUP_SPHINX`, but the original macro did use `OPAL_SUMMARY_ADD`.  Hence, I converted the `OPAL_SUMMARY` macros to `OAC` equivalents, too.

The one "somewhat unclean" thing about this is how `OAC_SETUP_SPHINX` can set `OAC_MAKEDIST_DISABLE` (if a suitable Sphinx is not found).  All it can do is set this environment variable; it can't really enforce anything about disabling `make dist`.  I.e., it's the caller's responsibility to check `$OAC_MAKEDIST_DISABLE` to see if `make dist` should be disabled.  This feels... a little unclean.  Got a better suggestion?

jsquyres commented:
@bwbarrett This PR is paired with a corresponding OMPI PR: https://github.com/open-mpi/ompi/pull/11190

jsquyres commented:
@bwbarrett Per our discussion, I added an `m4_pattern_forbid([OAC_])`.  It got a little messy where we emit `OAC_something` in some error messages -- I had to surround those sections with `m4_pattern_allow` / `m4_pattern_deny`.

bwbarrett commented on oac_summary.m4:
```diff
@@ -0,0 +1,98 @@
+# work properly with the current behavior, but would result in odd
+# errors if we delayed evaluation.
+#
+# As a final historical note, the initial version of SUMMARY_ADD was
```
This note can probably die now, since this is the first version of OAC_SUMMARY_ADD.

jsquyres replied:
Killed.

bwbarrett reviewed (approved):
I'm ok with this patch series.  However, I'll note that Open MPI and PRRTE/PMIX differ in where the summary information is output.  Open MPI (and this macro) uses stdout, while PRRTE/PMIX use stderr.

rhc54 commented:
> I'm ok with this patch series. However, I'll note that Open MPI and PRRTE/PMIX differ in where the summary information is output. Open MPI (and this macro) uses stdout, while PRRTE/PMIX use stderr.

Yeah, that's an ouch all right - it really helps me a lot to be able to redirect stdout away but still receive the summary. I suspect `oac_summary` is one I won't be able to use.

rhc54 commented:
Just wondering - can we add a flag to indicate output to stdout or stderr? I could use it then

jsquyres commented:
Pushed changes:

* Removed outdated comment (per Brian's note)
* Added optional $1 argument to `OAC_SUMMARY_PRINT`.  PMIx can invoke `OAC_SUMMARY_PRINT([stderr])` (which emits everything to stderr), whereas OMPI can invoke `OAC_SUMMARY_PRINT` (which emits everything to stdout).

rhc54 reviewed (approved):
Thanks!

bwbarrett commented on oac_summary.m4:
```diff
@@ -0,0 +1,96 @@
+dnl send the output to stdout.
+AC_DEFUN([OAC_SUMMARY_PRINT],[
+    OAC_VAR_SCOPE_PUSH([oac_summary_section oac_summary_section_name oac_sumary_fd])
+    AS_IF([test "$1" = "stderr"],
```
I'd prefer do this at autoconf time; this shouldn't be a cofigure time option.  Something like:

```
m4_define([oac_summary_print_fd], [m4_default([$1], [1])])
echo "foobar" >&oac_summary_print_fd
```

Note that I didn't actually test that, so some adjustments might be needed.  I'd also claim that the default should probably actually be AS_MESSAGE_FD instead of 1, so that the summary doesn't print when run with `configure -q`, but that's a bigger discussion.

rhc54 replied:
You mean as a command line option? I much prefer this method where I can just hard-code it with the desired behavior.  There's only one of "me" over here, so it doesn't make a lot of sense to add complexity for what is, after all, a pretty simple thing.

Note that I also don't want to change the output stream for _all_ AC output - I just want the old behavior where the summary comes out on stderr and everything else on stdout.

jsquyres replied:
I looked at `m4sh` yesterday and didn't find anything that would have allowed me to do it at autogen time; I neglected to look at the `m4` macros. `m4_default` looks perfect -- thanks for the suggestion.

I wanted to keep the ability to pass `stderr` instead of `2` (it's just a little more expressive to read `OAC_SUMMARY_PRINT([stderr])` than `OAC_SUMMARY_PRINT([2])`), so I ended up going with:

```
m4_define([oac_summary_print_fd], [m4_default(m4_if([$1], [stderr], [2]), [1])])
```

jsquyres replied:
@rhc No, there's no difference to what you'll be doing.  It's still effectively hard-coded in `config/pmix_configure.m4`; it's just done slightly differently (more Autoconf-ish).

rhc54 replied:
Got it - thx! I realized it myself earlier this morning as I looked more closely at what you guys were doing.

bwbarrett replied:
> I looked at `m4sh` yesterday and didn't find anything that would have allowed me to do it at autogen time; I neglected to look at the `m4` macros. `m4_default` looks perfect -- thanks for the suggestion.

Right; m4sh is all about making writing shell code easier, so everything in that section is really configure time code.  There are some weird exceptions, but for the most part, you need to look at the m4sugar section (https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.71/html_node/Programming-in-M4sugar.html) for autoconf-time macros.  In this case, it's all renames of m4 macros, but same deal.

> I wanted to keep the ability to pass `stderr` instead of `2` (it's just a little more expressive to read `OAC_SUMMARY_PRINT([stderr])` than `OAC_SUMMARY_PRINT([2])`), so I ended up going with:
> 
> ```
> m4_define([oac_summary_print_fd], [m4_default(m4_if([$1], [stderr], [2]), [AS_MESSAGE_FD])])
> ```

Ugh; I guess, but that's ugly.

jsquyres commented:
Updated per @bwbarrett's comments (also used `AS_MESSAGE_FD` instead of defaulting to fd 1).

bwbarrett commented on oac_summary.m4:
```diff
@@ -78,18 +78,16 @@ AC_DEFUN([OAC_SUMMARY_ADD],[
 dnl If $1 is "stderr", then send the output to stderr.  Otherwise,
```
This file's changes should be in the previous commit.

jsquyres replied:
🤬  Doh!

jsquyres commented:
@bwbarrett Fixed.

bwbarrett commented on oac_summary.m4:
```diff
@@ -0,0 +1,94 @@
+dnl send the output to stdout.
+AC_DEFUN([OAC_SUMMARY_PRINT],[
+    OAC_VAR_SCOPE_PUSH([oac_summary_section oac_summary_section_name])
+    m4_define([oac_summary_print_fd], [m4_default(m4_if([$1], [stderr], [2]), [1])])
```
Sorry, I didn't look at this closely enough.  I don't think this is right (or at least, it doesn't make sense).  If you want the valid values of the argument to be 'stderr' or not set, then test for those two cases.  I had provided code for the case where users could pass an arbitrary fd (which is why the use of the string stderr bugs me, because it's not an fd), in which case you need an else clause in your m4_if to pass $1.

As it stands now, passing anything other than stderr will result in the argument being ignored and printing to file descriptor 1.  I would say that we should either use the passed argument as an FD OR explicitly check for the three cases (not set, stderr, anything else) and abort if anything else is passed.

Finally, I thought you were going to use AS_MESSAGE_FD , not 1, as the default?

jsquyres commented:
@bwbarrett All code + commit messages should be correct now.
